### DISCUSSION TITLE
How to get user groups

### DISCUSSION THREAD

**Discussion by a1730** [authelia/authelia#6534] [Q&A] [answered]
I have an instance of Authelia running against LLDAP but for the life of me I cannot convince Authelia to tell me what groups the user belongs to even though it has the values.

Below is a transcript that shows the Authelia has the groups as seen on line 3. 
<code>
msg="Computed groups filter is (member=uid=authtest,ou=people,dc=example,dc=com)"
msg="Performing group search" attributes="[cn]" base_dn="ou=groups,dc=example,dc=com" deref=0 filter="(member=uid=authtest,ou=people,dc=example,dc=com)" mode=filter scope=2 </code>
msg="Profile details for user 'authtest' => groups: [group1 group2], emails [authtest@accounts.example.com]" method=POST path=/api/firstfactor remote_ip=10.128.0.25<code>
msg="Check authorization of subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)."
msg="ACL MISS Position 1 for subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)"
msg="ACL MISS Position 2 for subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)"
msg="ACL MISS Position 3 for subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)"
msg="ACL MISS Position 4 for subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)"
msg="ACL MISS Position 5 for subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)"
msg="ACL HIT Position 6 for subject username=authtest groups=group1,group2 ip=10.128.0.25 and object https://www.example.com/homepage (method GET)"
</code>

Some information:
Authelia version: "Authelia untagged-v4.37.5 (v4.38.0-beta3, 5d1fe75) is starting"

Post to `/api/firstfactor`
```json
{"status":"OK","data":{"redirect":"https://www.example.com/homepage"}}
```
Http POST to `/api/user/info` yields
```json
{
  "status": "OK",
  "data": {
    "display_name": "Authentication Shenanigan",
    "method": "totp",
    "has_totp": false,
    "has_webauthn": false,
    "has_duo": false
  }
}
```
The entity between the chair and the computer would like to know how to extract the information from Authelia.

Thank you for your help.

**Comment by james-d-elliott** [comment]
The API is intended to service the web frontend and as such doesn't need this information. What was the intent behind accessing groups?

**Comment by a1730** [comment]
We are replacing a home grown authentication system with Authelia. 

Our application relies on user group to show even the landing page because we filter out menu items that the user/group does not have access to. 

FWIW: We have been able to get this information from other authnz providers: Auth0, FusiaonAuth, Glewlwyd, Kratos,  ... They each surface the information in some form on /userinfo. However none of them comes close to Authelia in ease of deployment.

Thank you for making auth feel easy.

**Reply by james-d-elliott** [reply]
You can probably leverage the authz endpoints starting with `/api/authz/*` or the `/api/verify` endpoint, they have a response header named `Remote-Groups` which is a CSV list of groups. Let me know if that's insufficient.

**Reply by james-d-elliott** [reply]
Alternatively these headers can be injected as a request header to the backend (by copying them from the response into the request) along with the username using native authelia proxy integrations relatively easily.

**Reply by a1730** [reply]
We are using the api directly without a proxy.
I tested /api/verify and it seems to do the job. The battle continues ...

Thank you for the timely response.

**Comment by a1730** [comment]
I did a quick test with /api/verify with a logged in account and it works but with a  **lot of ceremony**, and another HTTP request to boot. 
The amount of information required is sub-optimal. Even with that, we were willing to go with /api/verify until we noticed that it is tagged as Legacy by the Authelia team. Furthermore, we are not planning to use Authelia for Authz. The application has fine grained authz built in requiring only who and the group affiliation.

For these reasons, and the fact that the required parameters for /api/verify have already been evaluated during login by Authelia - it just needs to surface the value, please consider a facility for providing user groups. It would simplify our life immensely and Authelia would satisfy our Authn requirements completely without friction. e.g something like the follow from _/api/user/info_.
```json
{
  "status": "OK",
  "data": {
    "display_name": "Authentication Shenanigan",
    "groups": "group1, group2", 
    "groupz": ["group1", "group2"], 
    "method": "totp",
    "has_totp": false,
    "has_webauthn": false,
    "has_duo": false
  }
}
```
Just to be clear, we made _/api/verify_ work but we believe that Authelia will be the bigger, better, stronger with the addition of groups to the response from _/api/user/info_.
Thank you for the swift response.

**Reply by james-d-elliott** [reply]
Realistically the solution I'd have suggested if the previous didn't work for you was OpenID Connect 1.0 as it is quite literally what it's designed for. It includes many measures to prevent MITM attacks unlike all of the other solutions.

**Comment by a1730** [comment]
In trying _/api/authz/authz-request_, we noticed that it wants the original request - this is sub-optimal for a page with a form where the original request is a GET, the intermediate request is POST, and final request (redirect) depends on the outcome of the POST.

We will stick to **OPTIONS _/api/verify_** for now, reusing the original values of X-Forwarded-Host and X-Forwarded-URI from the original authentication _/api/firstfactor_ request.

Thank you.

**Comment by a1730** [comment]
> Alternatively these headers can be injected as a request header to the backend (by copying them from the response into the request) along with the username using native authelia proxy integrations relatively easily.

I did not see these headers on the response for _/api/firstfactor_ nor on _/api/user/info_. Please tell me more or point me to documentation, please.

**Reply by james-d-elliott** [reply]
Each instance of authelia documents its API at the /api/ path.

**Comment by a1730** [comment]
We found a solution using LLDAP [scripting api](https://github.com/lldap/lldap/blob/main/docs/scripting.md). Each tool now does what is does best, Authelia handles the front end authentication, our application makes a side channel request to LLDAP for user information in response to which LLDAP responds with a json payload of user information. Life is good :tm:

Most of the information came from these places [GetUserDetails](https://github.com/lldap/lldap/blob/main/app/queries/get_user_details.graphql), [scripting](https://github.com/lldap/lldap/blob/main/docs/scripting.md), [here](https://github.com/lldap/lldap/issues/707#issuecomment-1871185517), [here](https://docs.google.com/document/d/1BdizgwX6Pvy5hqH8_HzSg3LCa66tQSDhrTsdhF80ySY/edit#heading=h.ht50h86g9sqq), and [lldap-cli](https://github.com/Zepmann/lldap-cli)

A big thank you to @james-d-elliott and @nitnelave from the LLDAP project for his support on lldap/lldap#789

**Comment by a1730** [comment]
> Realistically the solution I'd have suggested if the previous didn't work for you was OpenID Connect 1.0 as it is quite literally what it's designed for. It includes many measures to prevent MITM attacks unlike all of the other solutions.

You are short changing the hard work you have put into Authelia by recommending OIDC "flow" for something that plays very well into Authelia's strength.  The only thing OIDC has on Authelia, in my opinion, is decentralization. Sometimes, users just don't need nor want all the ceremony surrounding "standards" like OIDC. OIDC it is good for Google, Microsoft, Github e.t.c (I don't think Amazon.com does OIDC) but sometimes one just wants a secure handling and storage of identity configuration, and authentication in a **SMB**.  _Authelia does a fantastic work at this!_ Authelia offers a complete (rich set of configuration options including OIDC), secure, lightweight,  low ceremony solution for _login and authentication_. Kudos to you and the Authelia team for the design. You should compare Authelia to the Ory suite one of these days and I hope that you pat yourself on the back😄 The Ory suite is very good and  comprehensive but mind the moving parts ...

Identity management is another issue - Authelia does not have to do _everything._ Thinking about this request, I was unfair asking for identity information from Authelia. I should be asking the identity provider for that information - hey, I did not think it through. However, if I was using Authelia's file provider then, I would expect to get user information from Authlia😉 With that said, others have been doing _identity management_ for ages. I believe that Authelia should let the Database, OIDC, LDAP, IAM, CIAM ceremony providers deal with that concern. 

I wish to thank you for the time you have given me on this issue. **Authelia is very, very good at what it does.**  _The Single Sign-On Multi-Factor portal for web apps _.  We are sticking to Authelia, thank you.

Please keep up the good work. I wish you all a lof of success and we are rooting for you and the team.

Thank you again.

**Comment by james-d-elliott** [comment, marked as answer]
> Thinking about this request, I was unfair asking for identity information from Authelia. I should be asking the identity provider for that information - hey, I did not think it through.

Authelia provides group information in two distinct and relatively well supported ways.

If you want low ceremony options our proxy integration will work flawlessly in practically every situation. We provide easy to use response headers that can be easily copied/injected into an app for RBAC (username, email, groups, display name). All that's required is the intermediary request to the authorization endpoints (`/api/authz/*` or `/api/verify`) which is currently supported in all major webservers / proxies. 

For a more modern and secure approach which pushes a majority of the authentication and RBAC onto the protected application OpenID Connect 1.0 / OAuth 2.0 (we support and is effectively an API based flow which can be completely transparent to the user if desired) or SAML 2.0 (which we don't support yet) are quite easily the best bets. OpenID Connect 1.0 in particular has client support in every modern language and is very easy to implement. These methods also allow an informed third party to verify the information has not been tampered with in any way.

**Reply by a1730** [reply]
Thank you very much for the clarification.
My apologies if my previous comments ruffls a few feathers.  My  comment was meant to be a positive endorsement of Authelia to ensure that no one sells Authelia short. I repeat **Authelia is very, very good at what it does  _The Single Sign-On Multi-Factor portal for web apps_**  Check out the Github stars 🌟 

Thank you for a great project.
